File: src/main/java/com/bibireden/wizardex/WizardEX.java
package com.bibireden.wizardex;

import net.fabricmc.api.ModInitializer;
import net.minecraft.util.Identifier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.github.clevernucleus.dataattributes_dc.api.DataAttributesAPI;
import com.github.clevernucleus.playerex.api.EntityAttributeSupplier;
import com.github.clevernucleus.playerex.api.ExAPI;


public class WizardEX implements ModInitializer {
	public static final String ID = "wizardex";
	public static final Logger LOGGER = LoggerFactory.getLogger(ID);


	// Overall spell stat gain from int levels
	public static final Identifier ALL_SCHOOLS = new Identifier(ID, "spell_power_all");

	// Indvidual Spell Levels
	public static final Identifier FIRE_SCHOOL = new Identifier(ID, "spell_power_fire");
	public static final Identifier FROST_SCHOOL = new Identifier(ID, "spell_power_frost");
	public static final Identifier LIGHTNING_SCHOOL = new Identifier(ID, "spell_power_lightning");
	public static final Identifier SOUL_SCHOOL = new Identifier(ID, "spell_power_soul");
	public static final Identifier ARCANE_SCHOOL = new Identifier(ID, "spell_power_arcane");
	public static Identifier HEALING_SCHOOL = new Identifier(ID, "spell_power_healing");

	public static final Identifier CRIT_CHANCE_SP = new Identifier("spell_power:critical_chance");
	public static final Identifier CRIT_DAMAGE_SP = new Identifier("spell_power:critical_damage");

	@Override
	public void onInitialize() {
		LOGGER.debug("WizardEX is preparing magic!");
		registerRefundConditions();
	}

	private void registerRefundConditions() {
		ExAPI.registerRefundCondition((data, player) -> {
			var attribute = EntityAttributeSupplier.of(FIRE_SCHOOL);
			return DataAttributesAPI.ifPresent(player, attribute, 0.0D, value -> data.get(attribute));
		});
		ExAPI.registerRefundCondition((data, player) -> {
			var attribute = EntityAttributeSupplier.of(FROST_SCHOOL);
			return DataAttributesAPI.ifPresent(player, attribute, 0.0D, value -> data.get(attribute));
		});
		ExAPI.registerRefundCondition((data, player) -> {
			var attribute = EntityAttributeSupplier.of(ARCANE_SCHOOL);
			return DataAttributesAPI.ifPresent(player, attribute, 0.0D, value -> data.get(attribute));
		});
	}
}

File: src/main/java/com/bibireden/wizardex/WizardEX.kt
package com.bibireden.wizardex

import com.bibireden.data_attributes.api.attribute.StackingBehavior
import com.bibireden.data_attributes.api.factory.DefaultAttributeFactory
import com.bibireden.data_attributes.config.functions.AttributeFunction
import com.bibireden.data_attributes.data.EntityTypeData
import com.bibireden.playerex.api.PlayerEXAPI
import com.bibireden.playerex.api.attribute.PlayerEXAttributes
import com.bibireden.playerex.ext.id
import com.bibireden.wizardex.attributes.WizardEXAttributes
import net.fabricmc.api.ModInitializer
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.ai.attributes.RangedAttribute
import net.spell_power.SpellPowerMod
import net.spell_power.api.SpellPower
import net.spell_power.api.SpellSchool
import net.spell_power.api.SpellSchools
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import kotlin.reflect.KCallable
import kotlin.reflect.full.memberProperties

object WizardEX : ModInitializer {
    const val MOD_ID: String = "wizardex"

    @JvmField
    val SPELL_POWER_SCHOOLS: List<SpellSchool> = listOf(
        SpellSchools.FIRE,
        SpellSchools.FROST,
        SpellSchools.SOUL,
        SpellSchools.ARCANE,
        SpellSchools.HEALING,
        SpellSchools.LIGHTNING
    )
    @JvmField
    val SCHOOLS: List<RangedAttribute> = listOf(
        WizardEXAttributes.Schools.FIRE,
        WizardEXAttributes.Schools.FROST,
        WizardEXAttributes.Schools.SOUL,
        WizardEXAttributes.Schools.ARCANE,
        WizardEXAttributes.Schools.HEALING,
        WizardEXAttributes.Schools.LIGHTNING,
    )

    @JvmField
    val LOGGER: Logger = LoggerFactory.getLogger(MOD_ID)

    @JvmStatic
    fun id(str: String): ResourceLocation = ResourceLocation.tryBuild(MOD_ID, str)!!

    private fun registerRefundConditions() {
        SPELL_POWER_SCHOOLS.forEach { school -> PlayerEXAPI.registerRefundCondition { data, player -> data.get(school.attribute) } }
    }

    override fun onInitialize() {
        LOGGER.debug("WizardEX is preparing magic!")

        registerRefundConditions()

        DefaultAttributeFactory.registerFunctions(mapOf(
            PlayerEXAttributes.DEXTERITY.id to listOf(
                AttributeFunction(WizardEXAttributes.CRITICAL_DAMAGE.id, StackingBehavior.Add, 0.5)
            ),
            PlayerEXAttributes.INTELLIGENCE.id to SCHOOLS.map { AttributeFunction(it.id, StackingBehavior.Add, 0.25) },
            PlayerEXAttributes.LUCKINESS.id to listOf(
                AttributeFunction(WizardEXAttributes.CRITICAL_CHANCE.id, StackingBehavior.Add, 0.25)
            ),
            WizardEXAttributes.CRITICAL_CHANCE.id to listOf(
                AttributeFunction(ResourceLocation.tryBuild("spell_power", "critical_chance")!!, StackingBehavior.Add, 1.0)
            ),
            WizardEXAttributes.CRITICAL_DAMAGE.id to listOf(
                AttributeFunction(ResourceLocation.tryBuild("spell_power", "critical_damage")!!, StackingBehavior.Add, 1.0)
            ),

            WizardEXAttributes.Schools.FIRE.id to listOf(
                AttributeFunction(SpellSchools.FIRE.id, StackingBehavior.Add, 0.5)
            ),
            WizardEXAttributes.Schools.FROST.id to listOf(
                AttributeFunction(SpellSchools.FROST.id, StackingBehavior.Add, 0.5)
            ),
            WizardEXAttributes.Schools.LIGHTNING.id to listOf(
                AttributeFunction(SpellSchools.LIGHTNING.id, StackingBehavior.Add, 0.5)
            ),
            WizardEXAttributes.Schools.HEALING.id to listOf(
                AttributeFunction(SpellSchools.HEALING.id, StackingBehavior.Add, 0.5)
            ),
            WizardEXAttributes.Schools.ARCANE.id to listOf(
                AttributeFunction(SpellSchools.ARCANE.id, StackingBehavior.Add, 0.5)
            ),
            WizardEXAttributes.Schools.SOUL.id to listOf(
                AttributeFunction(SpellSchools.SOUL.id, StackingBehavior.Add, 0.5)
            ),
        ))

        DefaultAttributeFactory.registerEntityTypes(mapOf(
            ResourceLocation.tryBuild("minecraft", "player")!! to EntityTypeData(
                mutableMapOf(WizardEXAttributes.CRITICAL_CHANCE.id to 0.0, WizardEXAttributes.CRITICAL_DAMAGE.id to 0.0)
                    .also { it.putAll(SCHOOLS.map { s -> s.id to 0.0 }) }
            )
        ))
    }
}